### DISCUSSION TITLE
Strongswan interoperability with CISCO FlexVPN for INTENRAL_IP4_SUBNET

### DISCUSSION THREAD

**Discussion by lalitmishra87** [strongswan/strongswan#1251] [Q&A]
Hi,
	I am trying to interop CISCO FlexVPN with Strongswan for INTERNAL_IP4_SUBNET attribute. 

1. Strongswan successfully received INTERNAL_IP4_ADDRESS, INTERNAL_IP4_DNS and processed it successfully.
2.  But when Strongswan is receiving INTERNAL_IP4_SUBNET from CISCO, its complaining 
& not able to process received SUBNET.  Following are the logs.

```
	Aug 22 17:49:17 10[IKE] <TEST-PSK|1> processing INTERNAL_IP4_ADDRESS attribute
	Aug 22 17:49:17 10[IKE] <TEST-PSK|1> processing INTERNAL_IP4_DNS attribute
	Aug 22 17:49:17 10[IKE] <TEST-PSK|1> installing DNS server 192.168.123.129 to /etc/resolv.conf
	Aug 22 17:49:17 10[IKE] <TEST-PSK|1> processing INTERNAL_IP4_SUBNET attribute
	Aug 22 17:49:17 10[CFG] <TEST-PSK|1> handling INTERNAL_IP4_SUBNET attribute failed

```
**Objective**:

1. I am trying to achieve "IKEv2 dynamic routing". CISCO FlexVPN is sending 'routing information of its local network' over IKEv2 and Strongswan should be able to process it and should redirect traffic towards received routing subnet using ipsec tunnel.
2. In other terms, on my Strongswan rightsubnet would be dynamically received from CISCO as INTERNAL_IP4_SUBNET IKEv2 attribute and Strongswan should be able to use it as rightsubnet.

_**Note**_: I am using route based ipsec and VTI interfaces.
	
	
**My questions:**

1.  Does Strongswan apply & process received INTERNAL_IP4_SUBNET IKEv2 attribute from FlexVPN Server.
2. Is it supported in code or do I need to write some additional stuff to process it ?
4. Anyone achieved similar scenario ? If possible, Can I get some sample configuration ?
Any help is appreciated .

**Strongswan configuration:**
	
```

# ipsec.conf - strongSwan IPsec configuration file
config setup
#       charondebug="all"

# Add connections here.

conn %default
        ikelifetime=480m
        keylife=60m
        rekeymargin=3m
        keyingtries=1
        dpdaction=restart
conn TRO-PSK
        type=tunnel
        authby=secret
        keyexchange=ikev2
        ike=aes256-sha256-modp2048
        esp=aes256-sha256-modp2048
        left=192.168.1.101
        leftid=192.168.1.101
     ## leftsubnet=10.10.10.0/24
        rightsubnet=0.0.0.0/0
        leftfirewall=no
        right=192.168.1.100
        rightid=192.168.1.100
        leftsourceip=%config
        auto=start
        mark=101
```

**Comment by Thermi** [comment]
strongSwan doesn't know anything about VTIs or XFRMIs or does any dynamic routing, nor does anything with them. 

> Does Strongswan apply & process received INTERNAL_IP4_SUBNET IKEv2 attribute from FlexVPN Server.

No. Generally nothing with any INTERNAL_* attributes.

> Is it supported in code or do I need to write some additional stuff to process it ?

Need to write additional code to process it.

> Anyone achieved similar scenario ? If possible, Can I get some sample configuration ?
Any help is appreciated .

It's not generally possible right now. Doing the right thing with routes, or any host configuration depending on tunnel status is a difficult problem that we first need to figure out a good solution for outside of route management for the IKE_SAs.

I suppose one could write a plugin that handles these attributes and installs the routes, but the question is over what interface. strongSwan doesn't know about VTIs or XFRMIs, so I guess an additional setting would be needed to convey the information to the code doing this attribute handling.

**Reply by lalitmishra87** [reply]
Hi Thermi, Thank you for your response and help. 
Yes, right, I could not find handler for INTERNAL_IP4_SUBNET. Handling and processing is happening for INTENRAL_IP4_ADDRESS, INTERNAL_IP4_DNS in code but not for INTERNAL_IP4_SUBNET.

I am trying to play with VTI and additional routes to process received SUBNET information.

**Comment by scotia70** [comment]
I have FreeBSD/StrongSwan (charon) talking to Cisco IKEv2 with VTIs (+NAT-T).  I don't use the INTERNAL_IP4_SUBNET var.  The Cisco initiates the tunnels and then I run dynamic routing over the top.  I use it for hub & spoke where the router is the hub and the spokes are various FreeBSD machines.  If this is your use case let me know and I'll post the configs.
